aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,473 @@
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+
+#include <c10/util/intrusive_ptr.h>
+
+#include <iostream>
+#include <exception>
+
+namespace at {
+namespace autocast {
+
+namespace {
+// Imitate Apex and cache some of the casts to streamline parameter reuse.
+// Our heuristic is to cache fp16 casts of fp32 model weights (see cached_cast below).
+//
+// After discussion with @ezyang, the cache uses the following structure:
+// The key is the source tensor's TensorImpl*, a proxy for a Tensor uuid that's unchanged
+// across shallow copies.  The value is a tuple with a weakref to the source tensor's
+// TensorImpl as the first element and the casted tensor as the second element.
+//
+// The weakref keeps the source's TensorImpl from being deleted.  We need to because we're
+// using the source TensorImpl* as the key.  If it were deleted, another random Tensor could
+// be allocated whose TensorImpl* happened to have the same value.  This TensorImpl* would
+// then mistakenly hit in cache:  a rare, intermittent, unpredictable bug.
+//
+// I'm not using the weak_intrusive_ptr as the key because it's more difficult to compare
+// directly against incoming TensorImpl*s.
+using weakref_type = c10::weak_intrusive_ptr<TensorImpl, UndefinedTensorImpl>;
+using val_type = std::tuple<weakref_type, Tensor>;
+thread_local std::unordered_map<TensorImpl*, val_type> cached_casts;
```

Comment:
Conversely, this one is probably going trigger internal build failure without the macros. I'll import your diff and check on this.

Reply:
Thanks!